lib/black_iron/entities/entity.ex

```diff
@@ -0,0 +1,69 @@
+defmodule BlackIron.Entities.Entity do
+  use Ecto.Schema
+  import Ecto.Changeset
+
+  import PolymorphicEmbed
+
+  @derive {Jason.Encoder, only: [:pid, :data, :rev, :revisions, :deleted_at]}
+
+  schema "entities" do
+    field :pid, :string, autogenerate: {BlackIron.Utils, :gen_pid, []}
+    field :rev, :string
+    field :revisions, {:array, :string}
+    field :deleted_at, :naive_datetime
+
+    polymorphic_embeds_one(:data,
+      types: [
+        character: BlackIron.Characters.Character,
+        campaign: BlackIron.Campaigns.Campaign
+      ],
+      on_type_not_found: :raise,
+      on_replace: :update
+    )
+
+    timestamps(type: :utc_datetime)
+  end
+
+  @doc false
+  def deactivate_changeset(entity, attrs) do
+    entity
+    |> cast(attrs, [:deleted_at, :rev, :revisions])
+    |> validate_required([:deleted_at, :rev, :revisions])
+  end
+
+  @doc false
+  def changeset(entity, enstruct, attrs \\ %{}) do
+    attrs =
+      attrs
+      |> Map.put(
+        "data",
+        Map.get(attrs, "data", %{})
+        |> Map.put("__type__", to_string(get_polymorphic_type(__MODULE__, :data, enstruct)))
```

Comment:
Not sure precisely how to encode this-- but it feels like we should not let an entity's type to change once it is set. Feels like that could wreak havoc on things that reference it. What od you think?

Reply:
The way I'm defending against this right now is that every `<Entity>.change` function manually overrides `__type__`, so it can't be changed by user input.